### PR TITLE
Add Clifford circuit optimizer

### LINKED ISSUE #756: optimized_for_xmon() should cancel more CZ gates
These should optimize to an empty circuit.

Input 1:
```
0: ───@───────────@───
      │           │
1: ───@───@───@───@───
          │   │
2: ───────@───@───────
```
Current output:
```
0: ───@───@───
      │   │
1: ───@───@───
```

Input 2:
```
0: ───@───────@───────
      │       │
1: ───@───@───@───@───
          │       │
2: ───────@───────@───
```
Current output:
```
0: ───@───────@───────
      │       │
1: ───@───@───@───@───
          │       │
2: ───────@───────@───
```

### PR DESCRIPTION
**Pull request by cduck**
This optimizer could be used in optimized_for_xmon() to res​olve #756.

### CODE REVIEW

File: cirq/contrib/paulistring/clifford_optimize.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+                return len(current_string.pauli_string) != 1
+            if (isinstance(op, ops.GateOperation)
+                and isinstance(op.gate, ops.Rot11Gate)):
+                return not stop_at_cz
+            if (isinstance(op, PauliStringPhasor)
+                and len(op.qubits) == 1
+                and (op.pauli_string[op.qubits[0]]
+                     == current_string.pauli_string[op.qubits[0]])):
+                return 2
+            return False
```
Comment (Strilanc) [inline cirq/contrib/paulistring/clifford_optimize.py]: This passes typecheck? False isn't an int..

File: cirq/contrib/paulistring/clifford_optimize.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+        modified_op = string_op
+        furthest_op = string_op
+        furthest_i = start_i + 1
+        i = start_i
```
Comment (Strilanc) [inline cirq/contrib/paulistring/clifford_optimize.py]: Dead assignment

File: cirq/contrib/paulistring/clifford_optimize.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        for i in range(start_i+1, len(all_ops)):
```
Comment (Strilanc) [inline cirq/contrib/paulistring/clifford_optimize.py]: Kinda looks like start_i should just be one higher.

File: cirq/contrib/paulistring/clifford_optimize.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    all_ops = list(c_cliff.all_operations())
+
+    def find_merge_point(start_i: int,
+                         string_op: PauliStringPhasor,
+                         stop_at_cz: bool,
+                         ) -> Tuple[int, PauliStringPhasor]:
+        def continue_condition(op: ops.Operation,
+                               current_string: PauliStringPhasor,
+                               is_first: bool) -> int:
+            if (isinstance(op, ops.GateOperation)
```
Comment (Strilanc) [inline cirq/contrib/paulistring/clifford_optimize.py]: Add a docstring with the 0,1,2 meanings.

File: cirq/contrib/paulistring/clifford_optimize.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        return furthest_i, furthest_op
+
+    def try_merge_clifford(cliff_op: ops.GateOperation, start_i: int) -> bool:
+        orig_qubit, = cliff_op.qubits
+        remaining_cliff_gate = ops.CliffordGate.I
+        for pauli, quarter_turns in reversed(
+                cast(ops.CliffordGate, cliff_op.gate).decompose_rotation()):
+            trans = remaining_cliff_gate.transform(pauli)
+            pauli = trans.to
+            quarter_turns *= (1, -1)[trans.flip]
```
Comment (Strilanc) [inline cirq/contrib/paulistring/clifford_optimize.py]: Interesting way to do this.

File: cirq/contrib/paulistring/clifford_optimize.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+                                            part_cliff_gate)
+
+        if remaining_cliff_gate == ops.CliffordGate.I:
+            all_ops.pop(start_i)
+            return True
+        else:
+            all_ops[start_i] = remaining_cliff_gate(orig_qubit)
+            return False
+
+    i = 0
```
Comment (Strilanc) [inline cirq/contrib/paulistring/clifford_optimize.py]: Split these giant loops into separate functions.

File: cirq/contrib/paulistring/clifford_optimize.py (hunk 1 above)
Comment (cduck) [reply cirq/contrib/paulistring/clifford_optimize.py]:
```
>>> False + False
0
```

Comment (Strilanc) [reply cirq/contrib/paulistring/clifford_optimize.py]:
```
>>> +False
0
```

File: cirq/contrib/paulistring/clifford_optimize.py (hunk 3 above)
Comment (cduck) [reply cirq/contrib/paulistring/clifford_optimize.py]: It's consistent between the two helper functions.

File: cirq/contrib/paulistring/clifford_optimize.py (hunk 6 above)
Comment (cduck) [reply cirq/contrib/paulistring/clifford_optimize.py]: Done.

File: cirq/contrib/paulistring/clifford_optimize.py (hunk 4 above)
Comment (cduck) [reply cirq/contrib/paulistring/clifford_optimize.py]: Done.

File: cirq/contrib/paulistring/clifford_optimize.py (hunk 2 above)
Comment (cduck) [reply cirq/contrib/paulistring/clifford_optimize.py]: Fixed.

File: cirq/contrib/paulistring/clifford_optimize_test.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+    c_opt = clifford_optimized_circuit(c_orig)
+
+    cirq.testing.assert_allclose_up_to_global_phase(
+        c_orig.to_unitary_matrix(),
+        c_opt.to_unitary_matrix(qubits_that_should_be_present=(q0, q1)),
+        atol=1e-7,
+    )
+
+    print(c_opt)
```
Comment (Strilanc) [inline cirq/contrib/paulistring/clifford_optimize_test.py]: Debug lines

**Pull request merged**